src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,338 @@
+        //    {
+        //        return Array.Empty<T>();
+
+        //    }
+        //    //We use JsonObjects for the arrays so we advance to the first property in the object which is the name/ver of the first library
+        //    Read();
+
+        //    if (TokenType == JsonTokenType.EndObject)
+        //    {
+        //        return Array.Empty<T>();
+        //    }
+
+        //    var listObjects = new List<T>();
+        //    do
+        //    {
+        //        listObjects.Add(streamReaderConverter.Read(ref this));
+        //        //At this point we're looking at the EndObject token for the object, need to advance.
+        //        Read();
+        //    }
+        //    while (TokenType != JsonTokenType.EndObject);
+        //    return listObjects;
+        //}
+
+        //internal void ReadArrayOfObjects<T1, T2>(IList<T2> objectList, Utf8JsonStreamReaderConverter<T1> streamReaderConverter) where T1 : T2
+        //{
+        //    if (objectList is null)
+        //    {
+        //        return;
+        //    }
+
+        //    if (Read() && TokenType == JsonTokenType.StartArray)
+        //    {
+        //        while (Read() && TokenType != JsonTokenType.EndArray)
+        //        {
+        //            var convertedObject = streamReaderConverter.Read(ref this);
+        //            if (convertedObject != null)
+        //            {
+        //                objectList.Add(convertedObject);
+        //            }
+        //        }
+        //    }
+        //}
+
+        internal string ReadNextTokenAsString()
+        {
+            if (_complete)
+            {
+                throw new InvalidOperationException("Cannot read from completed Utf8JsonStreamReader");
+            }
+
+            if (Read())
+            {
+                return _reader.ReadTokenAsString();
+            }
+
+            return null;
+        }
+
+        internal string GetCurrentBufferAsString()
+        {
+            if (_complete)
+            {
+                throw new InvalidOperationException("Cannot read from completed Utf8JsonStreamReader");
+            }
+
+            return Encoding.UTF8.GetString(_buffer);
+        }
+
+        internal IList<string> ReadStringArrayAsIList(IList<string> strings = null)
+        {
+            if (_complete)
+            {
+                throw new InvalidOperationException("Cannot read from completed Utf8JsonStreamReader");
+            }
+
+            if (TokenType == JsonTokenType.StartArray)
+            {
+                while (Read() && TokenType != JsonTokenType.EndArray)
+                {
+                    string value = _reader.ReadTokenAsString();
+
+                    strings = strings ?? new List<string>();
+
+                    strings.Add(value);
+                }
+            }
+
+            return strings;
+        }
+
+        internal List<string> ReadStringArrayAsList(List<string> strings = null)
```

Comment:
https://github.com/NuGet/NuGet.Client/pull/5529/files#r1414992003
> This looks very similar ReadStringArrayAsList above, unify into the one method? - @davkean 

Yea, the struggle was when the class being deserialized is creating a new list on instantiation, some of them have an IList, others have List, and others have IReadonlyList.  

I added a comment to make sure I look into merging or changing the classes we were deserializing to make sure they have consistent list implementations.

Reply:
I went through where it's used and we have two cases. ReadStringArrayAsList will be called and we set the property in the object we're creating with the results.

The instances where the objects instantiate the list object have their property set as an IList, so I can't pass in an IList to ReadStringArrayAsList and return a list, without forcing a cast. If we change it to return a List then all the places that are setting it to a list will need a cast. 

I updated it so most of the logic lives in the IList function and the List one just casts the results.